File: algorithmic_toolbox/2nd_week/5_fibonacci_number_again/fibonacci_huge.cpp
#include <iostream>

long long find_cycle(long long m)
{
    long long previous = 0;
    long long current = 1;
    long long temp;

    for (long long i = 0; i < m * m; i++)
    {
        temp = current;
        current = (current + previous) % m;
        previous = temp;

        // Check for the start of a new cycle
        if (previous == 0 && current == 1)
            return i + 1;
    }

    return 0;
}

long long fibonacci_huge_fast(long long n, long long m)
{
    if (m == 0)
        return 0;

    long long cycle = find_cycle(m);
    n = n % cycle;

    long long previous = 0;
    long long current = 1;

    for (long long i = 0; i < n - 1; i++)
    {
        long long temp = current;
        current = (current + previous) % m;
        previous = temp;
    }

    return current;
}

int main()
{
    long long n, m;
    std::cin >> n >> m;
    std::cout << fibonacci_huge_fast(n, m) << '\n';
}

File: algorithmic_toolbox/2nd_week/5_fibonacci_number_again/fibonacci_huge.py
def find_cycle(m):
    previous, current = 0, 1
    for i in range(0, m * m):
        previous, current = current, (previous + current) % m
        if previous == 0 and current == 1:
            return i + 1


def fibonacci_huge_fast(n, m):
    pisano_period = find_cycle(m)
    n = n % pisano_period

    previous, current = 0, 1
    if n == 0:
        return 0
    elif n == 1:
        return 1

    for _ in range(n - 1):
        previous, current = current, (previous + current) % m

    return current


if __name__ == '__main__':
    n, m = map(int, input().split())
    print(fibonacci_huge_fast(n, m))
